File: solutions/elixir/pig-latin/1/lib/pig_latin.ex
defmodule PigLatin do
  @doc """
  Given a `phrase`, translate it a word at a time to Pig Latin.

  Words beginning with consonants should have the consonant moved to the end of
  the word, followed by "ay".

  Words beginning with vowels (aeiou) should have "ay" added to the end of the
  word.

  Some groups of letters are treated like consonants, including "ch", "qu",
  "squ", "th", "thr", and "sch".

  Some groups are treated like vowels, including "yt" and "xr".
  """
  @vowel_sounds ~r/^a|^e|^i|^o|^u/
  @consonant_sounds ~r/(?<consonants>^[b|c|d|f|g|h|j|k|l|m|n|p|q|r|s|t|v|w|x|z]*)/
  @sometimes_vowel_like ~r/^x|^y/
  @suffix "ay"

  @erlang_largest_64_bit_integer 2_305_843_009_213_693_951
  @arbitrary_word_end @erlang_largest_64_bit_integer

  @spec translate(phrase :: String.t()) :: String.t()
  def translate(phrase) do
    phrase
    |> String.split(" ")
    |> Enum.map(&translate_word/1)
    |> Enum.join(" ")
  end

  defp translate_word(word) do
    cond do
      does_word_begin_with_sometimes_vowel_like_letter?(word) ->
        handle_word_beginning_with_sometimes_vowel_like_letter(word)

      does_word_begin_with_vowel?(word) ->
        handle_word_beginning_with_vowel(word)

      does_word_begin_with_consonant?(word) ->
        handle_word_beginning_with_consonant(word)
    end
  end

  defp handle_word_beginning_with_sometimes_vowel_like_letter(word) do
    first_letter = String.first(word)
    ending = String.slice(word, 1..@arbitrary_word_end)

    capture = Regex.named_captures(@consonant_sounds, String.first(ending))
    immediate_consonants = capture["consonants"]

    if immediate_consonants !== "",
      do: add_suffix(word),
      else: add_suffix(ending <> first_letter)
  end

  def handle_word_beginning_with_vowel(word),
    do: add_suffix(word)

  defp handle_word_beginning_with_consonant(word) do
    %{"consonants" => start} = Regex.named_captures(@consonant_sounds, word)
    length = String.length(start)
    ending = String.slice(word, length..@arbitrary_word_end)

    if String.last(start) === "q" and
         String.first(ending) === "u",
       do: add_suffix(String.slice(ending, 1..@arbitrary_word_end) <> start <> "u"),
       else: add_suffix(ending <> start)
  end

  defp add_suffix(word),
    do: word <> @suffix

  defp does_word_begin_with_vowel?(word),
    do: Regex.match?(@vowel_sounds, word)

  defp does_word_begin_with_consonant?(word),
    do: Regex.match?(@consonant_sounds, word)

  defp does_word_begin_with_sometimes_vowel_like_letter?(word),
    do: Regex.match?(@sometimes_vowel_like, word)
end

File: solutions/scala/pig-latin/1/src/main/scala/PigLatin.scala
import scala.util.matching.Regex

object PigLatin {
  val VowelSounds: Regex = "(^[a|e|i|o|u]).*".r
  val ConstantSounds: Regex = "(^[b|c|ch|d|f|g|h|j|k|l|m|n|p|q|r|s|sch|t|th|thr|v|w|x|z]*).*".r
  val SometimesVowelLike: Regex = "(^[x|y]).*".r
  val suffix = "ay"

  def translate(english: String): String =
    english.split(" ")
      .map(this.translateWord)
      .mkString(" ")

  private def translateWord(word: String): String = {
    word match {
      case SometimesVowelLike(_*) =>
        this.handleWordBeginningWithSometimesVowelLikeLetter(word)

      case VowelSounds(_*) =>
        this.handleWordBeginningWithVowel(word)

      case ConstantSounds(_*) =>
        this.handleWordBeginningWithConstant(word)
    }
  }

  private def handleWordBeginningWithConstant(word: String): String = {
    val start: String = this.findFirstConstants(word)
    val length = start.length
    val (_, ending) = word.splitAt(length)

    if (
      start.last.toString == "q" &&
        ending.charAt(0).toString == "u"
    ) {
      val (_, ending2) = ending.splitAt(1)
      this.addSuffix(s"$ending2${start}u")
    }
    else {
      this.addSuffix(s"$ending$start")
    }
  }

  private def findFirstConstants(word: String): String = {
    ConstantSounds.findFirstMatchIn(word)
      .get
      .group(1)
  }

  private def handleWordBeginningWithVowel(word: String): String =
    this.addSuffix(word)

  private def addSuffix(word: String): String = s"$word$suffix"

  private def handleWordBeginningWithSometimesVowelLikeLetter(word: String): String = {
    val (firstLetter, ending) = word.splitAt(1)

    val capture = this.findFirstConstants(ending)
    if (capture != "") {
      this.addSuffix(word)
    }
    else {
      this.addSuffix(s"$ending$firstLetter")
    }
  }
}